crates/dock/tests/mutability_check.rs

```diff
@@ -0,0 +1,88 @@
+#[cfg(test)]
+mod tests {
+    use expect_test::expect_file;
+    use regex::Regex;
+
+    fn check(dir_name: &str) {
+        let mut out = Vec::new();
+        let mut err = Vec::new();
+
+        let main_file_contents =
+            std::fs::read_to_string(format!("tests/mutability_check/{dir_name}/main.nail"))
+                .unwrap();
+        // //---stdoutで区切ることで、標準出力と標準エラー出力をテストする
+        // //---stderrは必ず//---stdoutより後に書くことが前提
+        let (main_contents, _) = main_file_contents
+            .split_once("//---stdout")
+            .unwrap_or((&main_file_contents, ""));
+        let mut main_contents = main_contents.trim().to_string();
+
+        match dock::execute(
+            dock::NailExecutablePath::RootFile(
+                std::env::current_dir()
+                    .unwrap()
+                    .join(format!("tests/mutability_check/{dir_name}/main.nail")),
+            ),
+            &mut out,
+            &mut err,
+            false,
+        ) {
+            Ok(_) => (),
+            Err(e) => match e {
+                // 型チェックのエラーはExecutionError::Nailで返るため何もしない
+                // それ以外は型チェック以外のエラーなのでpanicしてテストを失敗させる
+                dock::ExecutionError::Nail => (),
+                dock::ExecutionError::Io(e) => panic!("io error: {e}"),
+                dock::ExecutionError::InvalidRootNailFilePath(e) => {
+                    panic!("invalid root nail file path: {e}")
+                }
+            },
+        }
```

Comment:
**Optimization Issue**: The execution error handling within the 'dock::execute' function is not optimized for different types of errors, potentially leading to unnecessary panics and not leveraging Rust's error handling capabilities effectively. <br> **Fix**: Refactor the error handling to properly differentiate between recoverable and non-recoverable errors, and handle them appropriately without using panics for recoverable errors. <br> **Code Suggestion**: 
 ```
 -            Err(e) => match e {
 -                // 型チェックのエラーはExecutionError::Nailで返るため何もしない
 -                // それ以外は型チェック以外のエラーなのでpanicしてテストを失敗させる
 -                dock::ExecutionError::Nail => (),
 -                dock::ExecutionError::Io(e) => panic!("io error: {e}"),
 -                dock::ExecutionError::InvalidRootNailFilePath(e) => {
 -                    panic!("invalid root nail file path: {e}")
 -                }
 -            },
 +            Err(e) => {
 +                use dock::ExecutionError;
 +                match e {
 +                    ExecutionError::Nail => log::error!("Nail execution error occurred"),
 +                    ExecutionError::Io(e) => log::warn!("IO error: {e}"),
 +                    ExecutionError::InvalidRootNailFilePath(e) => log::error!("Invalid path: {e}"),
 +                }
 +            }
 ```